core/git_command.py

```diff
@@ -124,7 +125,7 @@ def raise_error(msg):
                                  cwd=working_dir or self.repo_path,
                                  env=environ,
                                  startupinfo=startupinfo)
-            stdout, stderr = p.communicate(stdin.encode(encoding="UTF-8") if stdin else None)
+            stdout, stderr = p.communicate((stdin.encode(encoding="UTF-8") if encode else stdin) if stdin else None)
```

Comment:
Pushing binary file content to git works fine for me here.

One question: Why is utf-8 enforced in text mode? This is the default encoding of git, but as you push file content here, it might be more reliable to use the value from git config `i18n.filesencoding` if available to decide how to encode stdin and fall back to utf-8 if setting is not present?